File: Structural/Decorator/Python/Bike.py
import abc

class Vehicle(metaclass = abc.ABCMeta):
    @abc.abstractmethod
    def build(self):
        return

class Bicycle(Vehicle):
    def build(self):
        print("Bicycle")

class VehicleDecorator(Vehicle):
    def __init__(self, vehicle):
        self.vehicle = vehicle

    def build(self):
        self.vehicle.build()

        
class VehicleWithEngine(VehicleDecorator):
    def __init__(self, vehicle, enginePower):
        super().__init__(vehicle)
        self.enginePower = enginePower

    def build(self):
        super().build()
        print("With {0}hp engine".format(self.enginePower))

bicycle = Bicycle()
moped = VehicleWithEngine(bicycle, 10)
moped.build()

File: Structural/Decorator/JavaScript/Bike.js
class Vehicle
{
    build()
    {
        throw new Error("Abstract method!");
    }
}

class Bicycle extends Vehicle
{
    build()
    {
        console.log("Bicycle");
    };
}

class VehicleDecorator extends Vehicle
{
    constructor(vehicle)
    {
        super();
        this.vehicle = vehicle;
    }
    
    build()
    {
        this.vehicle.build();
    }
}

class VehicleWithEngine extends VehicleDecorator
{
    constructor (vehicle, enginePower)
    {
        super(vehicle);
        this.enginePower = enginePower;
    }
   
    build()
    {
        super.build();
        console.log("With ".concat(this.enginePower.toString()).concat("hp engine"));
    }
}

// Test

var bicycle = new Bicycle();
var moped = new VehicleWithEngine(bicycle, 10);
moped.build();